src/validate.rs

```diff
@@ -1,39 +1,36 @@
-use cketh_common::eth_rpc::ValidationError;
 use ic_cdk::api::management_canister::http_request::HttpHeader;
 
 use crate::{
     constants::{CONTENT_TYPE_HEADER, SERVICE_HOSTS_BLOCKLIST},
     util::hostname_from_url,
 };
 
-pub fn validate_hostname(hostname: &str) -> Result<(), ValidationError> {
+pub fn validate_hostname(hostname: &str) -> Result<(), &'static str> {
     if SERVICE_HOSTS_BLOCKLIST.contains(&hostname) {
-        Err(ValidationError::HostNotAllowed(hostname.to_string()))
+        Err("Hostname not allowed")
     } else {
         Ok(())
     }
 }
 
-pub fn validate_url_pattern(url_pattern: &str) -> Result<(), ValidationError> {
-    validate_hostname(
-        &hostname_from_url(url_pattern).ok_or(ValidationError::CredentialPathNotAllowed)?,
-    )
+pub fn validate_url_pattern(url_pattern: &str) -> Result<(), &'static str> {
+    validate_hostname(&hostname_from_url(url_pattern).ok_or("Invalid hostname in URL")?)
 }
 
-pub fn validate_header_patterns(header_patterns: &[HttpHeader]) -> Result<(), ValidationError> {
+pub fn validate_header_patterns(header_patterns: &[HttpHeader]) -> Result<(), &'static str> {
     if header_patterns
         .iter()
         .any(|HttpHeader { name, .. }| name == CONTENT_TYPE_HEADER)
     {
-        Err(ValidationError::CredentialHeaderNotAllowed) // TODO: rename to `HeaderNotAllowed`
+        Err("Invalid header name")
     } else {
         Ok(())
     }
 }
 
-pub fn validate_api_key(api_key: &str) -> Result<(), ValidationError> {
+pub fn validate_api_key(api_key: &str) -> Result<(), &'static str> {
     if api_key.contains(['.', '/', '?', '&']) {
```

Comment:
For validation, typically whitelist is always safer than blacklist. Note that there are many more [characters](https://support.exactonline.com/community/s/knowledge-base#All-All-DNO-Content-urlcharacters) that should be excluded, but for example `.` is not one of them. It's for example typical to encounter such a character if the key is a JWT token. I would do the following:

1. reject any string whose length is above say 200.
2. assert the string only contains `Alphanumeric [0-9a-zA-Z], special characters $-_.+!*'(),`. Alternatively, use a special purpose crate for this, such as https://crates.io/crates/urlencoding

Reply:
Thanks for updating the validation logic, seems much better to me now!

> String only contains characters in 0123456789ABCDEFGHIJKLMNOPQRTSUVWXYZabcdefghijklmnopqrstuvwxyz$-_.+!*; removed ' and parentheses since these seem least likely to be used in a real API key and most likely to be used for some sort of injection attack if we use the API keys in other ways in the future.

Yes I think that's a good compromise, it's something we can always adapt if necessary.